examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,444 @@
+    /// Witness (sender_account) that registered above file_hash.
+    witness:   AccountAddress,
+    /// Timestamp when this record was added to this smart contract.
+    timestamp: Timestamp,
+}
+
+/// The FileAddedEvent is logged when a new file is added.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct FileAddedEvent {
+    /// Hash of the file to be registered.
+    file_hash: String,
+}
+
+/// Init function that creates this eSealing smart contract.
+#[init(contract = "eSealing", event = "Event")]
+fn contract_init<S: HasStateApi>(
+    _ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+) -> InitResult<State<S>> {
+    Ok(State::new(state_builder))
+}
+
+/// Register a new file.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - If file hash has already been registered by the same witness
+///   (sender_account).
+/// - If a smart contract tries to register the file hash.
+#[receive(
+    contract = "eSealing",
+    name = "registerFile",
+    parameter = "String",
+    error = "ContractError",
+    mutable,
+    enable_logger
+)]
+fn register_file<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> Result<(), ContractError> {
+    // Ensure that only accounts can register a file
+    let sender_account = match ctx.sender() {
+        Address::Contract(_) => bail!(ContractError::OnlyAccount),
+        Address::Account(account_address) => account_address,
+    };
+
+    let file_hash: String = ctx.parameter_cursor().get()?;
+
+    let (state, builder) = host.state_and_builder();
+
+    // Ensure that every account can only register each file hash once.
+    ensure!(!state.witness_exists(&file_hash, sender_account), ContractError::AlreadyRecorded);
+
+    let slot_time = ctx.metadata().slot_time();
+
+    let exists = state.file_exists(&file_hash);
+
+    match exists {
+        true => {
+            // File exists => Add record
+            state.add_record(&file_hash, sender_account, slot_time);
+            logger.log(&Event::WitnessAdded(WitnessAddedEvent {
+                file_hash,
+                witness: sender_account,
+                timestamp: slot_time,
+            }))?;
+        }
+        false => {
+            // File does not exist => Add file and add record
+            state.add_file(&file_hash, builder);
+            logger.log(&Event::FileAdded(FileAddedEvent {
+                file_hash: file_hash.clone(),
+            }))?;
+
+            state.add_record(&file_hash, sender_account, slot_time);
+            logger.log(&Event::WitnessAdded(WitnessAddedEvent {
+                file_hash,
+                witness: sender_account,
+                timestamp: slot_time,
+            }))?;
+        }
+    }
+
+    Result::Ok(())
+}
+
+/// Get all records about a specific file hash.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+#[receive(
```

Comment:
This does not scale that well.
I am not sure of the requirements for this contract, but since anyone can register as a witness, the list of all witnesses does not seem that interesting.
Would a query for whether a given account is a witness of a given file suffice? It could return the timestamp as well when there is a witness.